File: platform/polyglot/confluence_api/py/src/confluence_api/services/color_scheme_service.py
"""
Color scheme service for Confluence Data Center REST API v9.2.3.

Provides operations for managing global and space-level color schemes
including retrieval, updates, and resets to default.
"""

from __future__ import annotations

from typing import Any

from confluence_api.logger import create_logger

log = create_logger("confluence_api", __file__)


class ColorSchemeService:
    """Service for Confluence color scheme operations."""

    def __init__(self, client: Any) -> None:
        self._client = client

    # ── Global Color Scheme ──────────────────────────────────────────────

    def get_default_color_scheme(self) -> dict:
        """Retrieve the system default color scheme."""
        log.debug("get_default_color_scheme called")
        result = self._client.get("settings/lookandfeel/default")
        log.info("get_default_color_scheme succeeded")
        return result

    def get_global_color_scheme(self) -> dict:
        """Retrieve the current global (custom) color scheme."""
        log.debug("get_global_color_scheme called")
        result = self._client.get("settings/lookandfeel/custom")
        log.info("get_global_color_scheme succeeded")
        return result

    def update_global_color_scheme(self, data: dict) -> dict:
        """
        Update the global color scheme.

        Args:
            data: Color scheme configuration dict containing color values.

        Returns:
            Updated color scheme dict.
        """
        log.debug("update_global_color_scheme called")
        result = self._client.put("settings/lookandfeel/custom", json_data=data)
        log.info("update_global_color_scheme succeeded")
        return result

    def reset_global_color_scheme(self) -> dict:
        """Reset the global color scheme to the system default."""
        log.debug("reset_global_color_scheme called")
        result = self._client.delete("settings/lookandfeel/custom")
        log.info("reset_global_color_scheme succeeded")
        return result

    # ── Space Color Scheme ───────────────────────────────────────────────

    def get_space_color_scheme_type(self, space_key: str) -> dict:
        """
        Retrieve the color scheme type for a space (default or custom).

        Args:
            space_key: The space key.

        Returns:
            Dict indicating whether the space uses default or custom scheme.
        """
        log.debug("get_space_color_scheme_type called", {"space_key": space_key})
        result = self._client.get(f"settings/lookandfeel/space/{space_key}/type")
        log.info("get_space_color_scheme_type succeeded", {"space_key": space_key})
        return result

    def set_space_color_scheme_type(self, space_key: str, data: dict) -> dict:
        """
        Set the color scheme type for a space (default or custom).

        Args:
            space_key: The space key.
            data: Dict with the type setting (e.g. {"colorSchemeType": "custom"}).

        Returns:
            Updated type setting dict.
        """
        log.debug("set_space_color_scheme_type called", {"space_key": space_key})
        result = self._client.put(
            f"settings/lookandfeel/space/{space_key}/type", json_data=data
        )
        log.info("set_space_color_scheme_type succeeded", {"space_key": space_key})
        return result

    def get_space_color_scheme(self, space_key: str) -> dict:
        """Retrieve the custom color scheme for a space."""
        log.debug("get_space_color_scheme called", {"space_key": space_key})
        result = self._client.get(f"settings/lookandfeel/space/{space_key}/custom")
        log.info("get_space_color_scheme succeeded", {"space_key": space_key})
        return result

    def update_space_color_scheme(self, space_key: str, data: dict) -> dict:
        """
        Update the custom color scheme for a space.

        Args:
            space_key: The space key.
            data: Color scheme configuration dict containing color values.

        Returns:
            Updated color scheme dict.
        """
        log.debug("update_space_color_scheme called", {"space_key": space_key})
        result = self._client.put(
            f"settings/lookandfeel/space/{space_key}/custom", json_data=data
        )
        log.info("update_space_color_scheme succeeded", {"space_key": space_key})
        return result

    def reset_space_color_scheme(self, space_key: str) -> dict:
        """Reset the space color scheme back to the global default."""
        log.debug("reset_space_color_scheme called", {"space_key": space_key})
        result = self._client.delete(
            f"settings/lookandfeel/space/{space_key}/custom"
        )
        log.info("reset_space_color_scheme succeeded", {"space_key": space_key})
        return result


class AsyncColorSchemeService:
    """Async service for Confluence color scheme operations."""

    def __init__(self, client: Any) -> None:
        self._client = client

    # ── Global Color Scheme ──────────────────────────────────────────────

    async def get_default_color_scheme(self) -> dict:
        """Retrieve the system default color scheme."""
        log.debug("get_default_color_scheme called")
        result = await self._client.get("settings/lookandfeel/default")
        log.info("get_default_color_scheme succeeded")
        return result

    async def get_global_color_scheme(self) -> dict:
        """Retrieve the current global (custom) color scheme."""
        log.debug("get_global_color_scheme called")
        result = await self._client.get("settings/lookandfeel/custom")
        log.info("get_global_color_scheme succeeded")
        return result

    async def update_global_color_scheme(self, data: dict) -> dict:
        """Update the global color scheme."""
        log.debug("update_global_color_scheme called")
        result = await self._client.put("settings/lookandfeel/custom", json_data=data)
        log.info("update_global_color_scheme succeeded")
        return result

    async def reset_global_color_scheme(self) -> dict:
        """Reset the global color scheme to the system default."""
        log.debug("reset_global_color_scheme called")
        result = await self._client.delete("settings/lookandfeel/custom")
        log.info("reset_global_color_scheme succeeded")
        return result

    # ── Space Color Scheme ───────────────────────────────────────────────

    async def get_space_color_scheme_type(self, space_key: str) -> dict:
        """Retrieve the color scheme type for a space (default or custom)."""
        log.debug("get_space_color_scheme_type called", {"space_key": space_key})
        result = await self._client.get(f"settings/lookandfeel/space/{space_key}/type")
        log.info("get_space_color_scheme_type succeeded", {"space_key": space_key})
        return result

    async def set_space_color_scheme_type(self, space_key: str, data: dict) -> dict:
        """Set the color scheme type for a space (default or custom)."""
        log.debug("set_space_color_scheme_type called", {"space_key": space_key})
        result = await self._client.put(
            f"settings/lookandfeel/space/{space_key}/type", json_data=data
        )
        log.info("set_space_color_scheme_type succeeded", {"space_key": space_key})
        return result

    async def get_space_color_scheme(self, space_key: str) -> dict:
        """Retrieve the custom color scheme for a space."""
        log.debug("get_space_color_scheme called", {"space_key": space_key})
        result = await self._client.get(f"settings/lookandfeel/space/{space_key}/custom")
        log.info("get_space_color_scheme succeeded", {"space_key": space_key})
        return result

    async def update_space_color_scheme(self, space_key: str, data: dict) -> dict:
        """Update the custom color scheme for a space."""
        log.debug("update_space_color_scheme called", {"space_key": space_key})
        result = await self._client.put(
            f"settings/lookandfeel/space/{space_key}/custom", json_data=data
        )
        log.info("update_space_color_scheme succeeded", {"space_key": space_key})
        return result

    async def reset_space_color_scheme(self, space_key: str) -> dict:
        """Reset the space color scheme back to the global default."""
        log.debug("reset_space_color_scheme called", {"space_key": space_key})
        result = await self._client.delete(
            f"settings/lookandfeel/space/{space_key}/custom"
        )
        log.info("reset_space_color_scheme succeeded", {"space_key": space_key})
        return result

File: platform/polyglot/confluence_api/mjs/src/services/color-scheme-service.mjs
/**
 * @module services/color-scheme-service
 * @description Service for Confluence Color Scheme (Look and Feel) REST API operations.
 *
 * Manages the visual theme configuration at global and space levels.
 * Supports reading default, global, and space-specific color schemes,
 * updating them, and resetting to defaults.
 *
 * Confluence Data Center REST API v9.2.3.
 */

import { createLogger } from '../logger.mjs';

const log = createLogger('confluence_api', import.meta.url);

/**
 * Service class for Confluence color scheme / look-and-feel operations.
 */
export class ColorSchemeService {
  /**
   * @param {import('../client/FetchClient.mjs').FetchClient} client
   *   The Confluence fetch client instance.
   */
  constructor(client) {
    /** @private */
    this._client = client;
  }

  // ---------------------------------------------------------------------------
  // Global Color Scheme
  // ---------------------------------------------------------------------------

  /**
   * Get the default (built-in) color scheme.
   *
   * Returns the base color scheme that ships with Confluence before
   * any customizations are applied.
   *
   * @returns {Promise<Object>} Default color scheme object.
   */
  async getDefaultColorScheme() {
    log.debug('getDefaultColorScheme called');
    try {
      const result = await this._client.get('settings/lookandfeel/default');
      log.info('getDefaultColorScheme succeeded');
      return result;
    } catch (error) {
      log.error('getDefaultColorScheme failed', { message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Get the current global color scheme.
   *
   * Returns the color scheme currently applied to the Confluence instance
   * at the global level.
   *
   * @returns {Promise<Object>} Global color scheme object.
   */
  async getGlobalColorScheme() {
    log.debug('getGlobalColorScheme called');
    try {
      const result = await this._client.get('settings/lookandfeel');
      log.info('getGlobalColorScheme succeeded');
      return result;
    } catch (error) {
      log.error('getGlobalColorScheme failed', { message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Update the global color scheme.
   *
   * @param {Object} data - Color scheme update payload.
   * @param {Object} [data.headings] - Heading color configuration.
   * @param {Object} [data.links] - Link color configuration.
   * @param {Object} [data.menus] - Menu color configuration.
   * @param {Object} [data.header] - Header color configuration.
   * @param {Object} [data.content] - Content area color configuration.
   * @param {Object} [data.bordersAndDividers] - Borders color configuration.
   * @returns {Promise<Object>} The updated global color scheme.
   */
  async updateGlobalColorScheme(data) {
    log.debug('updateGlobalColorScheme called');
    try {
      const result = await this._client.put('settings/lookandfeel', data);
      log.info('updateGlobalColorScheme succeeded');
      return result;
    } catch (error) {
      log.error('updateGlobalColorScheme failed', { message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Reset the global color scheme to the default.
   *
   * Removes all global customizations and reverts to the built-in theme.
   *
   * @returns {Promise<Object>} The reset color scheme.
   */
  async resetGlobalColorScheme() {
    log.debug('resetGlobalColorScheme called');
    try {
      const result = await this._client.delete('settings/lookandfeel');
      log.info('resetGlobalColorScheme succeeded');
      return result;
    } catch (error) {
      log.error('resetGlobalColorScheme failed', { message: error.message, status: error.status });
      throw error;
    }
  }

  // ---------------------------------------------------------------------------
  // Space Color Scheme
  // ---------------------------------------------------------------------------

  /**
   * Get the color scheme type for a space.
   *
   * Returns whether the space uses the global theme or a custom theme.
   *
   * @param {string} spaceKey - The space key.
   * @returns {Promise<Object>} Space color scheme type (global or custom).
   */
  async getSpaceColorSchemeType(spaceKey) {
    log.debug('getSpaceColorSchemeType called', { spaceKey });
    try {
      const result = await this._client.get(`settings/lookandfeel/custom/${spaceKey}/type`);
      log.info('getSpaceColorSchemeType succeeded', { spaceKey });
      return result;
    } catch (error) {
      log.error('getSpaceColorSchemeType failed', { spaceKey, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Set the color scheme type for a space.
   *
   * Switches a space between using the global theme and a custom theme.
   *
   * @param {string} spaceKey - The space key.
   * @param {Object} data - Type configuration payload.
   * @param {string} data.themeType - Theme type ("global" or "custom").
   * @returns {Promise<Object>} The updated space color scheme type.
   */
  async setSpaceColorSchemeType(spaceKey, data) {
    log.debug('setSpaceColorSchemeType called', { spaceKey, themeType: data?.themeType });
    try {
      const result = await this._client.put(`settings/lookandfeel/custom/${spaceKey}/type`, data);
      log.info('setSpaceColorSchemeType succeeded', { spaceKey });
      return result;
    } catch (error) {
      log.error('setSpaceColorSchemeType failed', { spaceKey, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Get the custom color scheme for a space.
   *
   * @param {string} spaceKey - The space key.
   * @returns {Promise<Object>} Space color scheme object.
   */
  async getSpaceColorScheme(spaceKey) {
    log.debug('getSpaceColorScheme called', { spaceKey });
    try {
      const result = await this._client.get(`settings/lookandfeel/custom/${spaceKey}`);
      log.info('getSpaceColorScheme succeeded', { spaceKey });
      return result;
    } catch (error) {
      log.error('getSpaceColorScheme failed', { spaceKey, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Update the custom color scheme for a space.
   *
   * @param {string} spaceKey - The space key.
   * @param {Object} data - Color scheme update payload.
   * @returns {Promise<Object>} The updated space color scheme.
   */
  async updateSpaceColorScheme(spaceKey, data) {
    log.debug('updateSpaceColorScheme called', { spaceKey });
    try {
      const result = await this._client.put(`settings/lookandfeel/custom/${spaceKey}`, data);
      log.info('updateSpaceColorScheme succeeded', { spaceKey });
      return result;
    } catch (error) {
      log.error('updateSpaceColorScheme failed', { spaceKey, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Reset a space's color scheme to the global default.
   *
   * Removes the space-level customization and reverts to the global theme.
   *
   * @param {string} spaceKey - The space key.
   * @returns {Promise<Object>} The reset result.
   */
  async resetSpaceColorScheme(spaceKey) {
    log.debug('resetSpaceColorScheme called', { spaceKey });
    try {
      const result = await this._client.delete(`settings/lookandfeel/custom/${spaceKey}`);
      log.info('resetSpaceColorScheme succeeded', { spaceKey });
      return result;
    } catch (error) {
      log.error('resetSpaceColorScheme failed', { spaceKey, message: error.message, status: error.status });
      throw error;
    }
  }
}
